data_structures/rb_tree2.cpp

```diff
@@ -0,0 +1,1490 @@
+                return node->parent->right;
+            }
+            else {
+                // no sibling
+                throw invalid_argument("Invalid node passed.");
+            }
+        }
+
+    private:
+        unsigned long long int numItems;
+        RBTNode* root;
+        vector<int> nodes_order_added;
+
+        /**
+         * @brief ToInfixString (private impl) -- print left, print root, print right
+         * 
+         * @param node 
+         * @return string 
+         */
+        string ToInfixString(struct RBTNode* node) const {
+            ostringstream nodeOSS;
+
+            if (node != nullptr) {
+                // print left (recursion)
+                nodeOSS << ToInfixString(node->left);
+
+                // print node's color and value
+                nodeOSS << " ";
+                if (node->color == COLOR_BLACK) {
+                    nodeOSS << 'B';
+                } else {
+                    nodeOSS << 'R';
+                }
+                nodeOSS << node->val << " ";
+
+                // print right (recursion)
+                nodeOSS << ToInfixString(node->right);
+            }
+            // return string once null node reached (base case)
+            return nodeOSS.str();
+        };
+
+        /**
+         * @brief ToPrefixString (private impl) -- prints node, then left, then right
+         * 
+         * @param node 
+         * @return string 
+         */
+        string ToPrefixString(struct RBTNode* node) const {
+            ostringstream nodeOSS;
+            if (node != nullptr) {
+                // print node's color and value
+                nodeOSS << " ";
+                if (node->color == COLOR_BLACK) {
+                    nodeOSS << 'B';
+                } else {
+                    nodeOSS << 'R';
+                }
+                nodeOSS << node->val << " ";
+
+                // print left then right (recursively)
+                nodeOSS << ToPrefixString(node->left);
+                nodeOSS << ToPrefixString(node->right);
+            }
+            return nodeOSS.str();
+        };
+
+        /**
+         * @brief ToPostfixString (private impl) -- prints left, then right, then node
+         * 
+         * @param node 
+         * @return string 
+         */
+        string ToPostfixString(struct RBTNode* node) const {
+            ostringstream nodeOSS;
+            if (node != nullptr) {
+                // print left then right (recursively)
+                nodeOSS << ToPostfixString(node->left);
+                nodeOSS << ToPostfixString(node->right);
+
+                // print node's color and value
+                nodeOSS << " ";
+                if (node->color == COLOR_BLACK) {
+                    nodeOSS << 'B';
+                } else {
+                    nodeOSS << 'R';
+                }
+                nodeOSS << node->val << " ";
+            }
+            return nodeOSS.str();
+        };
+};
+
+void TestSimpleConstructor(){
```

Comment:
The test functions should be documented as well. Something simple like this should work.

```c++
/**
 * @brief Self-test by using a simple constructor
 * @returns void
 */
```

This is so that Doxygen detects the functions as well.